src/dbus/dbus_thread_object.cpp

```diff
@@ -0,0 +1,559 @@
+/*
+ *    Copyright (c) 2020, The OpenThread Authors.
+ *    All rights reserved.
+ *
+ *    Redistribution and use in source and binary forms, with or without
+ *    modification, are permitted provided that the following conditions are met:
+ *    1. Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *    2. Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *    3. Neither the name of the copyright holder nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ *    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *    ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *    CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *    SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *    INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *    CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *    ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *    POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <byteswap.h>
+#include <string.h>
+
+#include "openthread/border_router.h"
+#include "openthread/instance.h"
+#include "openthread/joiner.h"
+
+#include "dbus/constants.hpp"
+#include "dbus/dbus_agent.hpp"
+#include "dbus/dbus_thread_object.hpp"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+
+static std::string GetDeviceRoleName(otDeviceRole aRole)
+{
+    std::string roleName;
+
+    switch (aRole)
+    {
+    case OT_DEVICE_ROLE_DISABLED:
+        roleName = OTBR_DISABLED_ROLE_NAME;
+        break;
+    case OT_DEVICE_ROLE_DETACHED:
+        roleName = OTBR_DETACHED_ROLE_NAME;
+        break;
+    case OT_DEVICE_ROLE_CHILD:
+        roleName = OTBR_CHILD_ROLE_NAME;
+        break;
+    case OT_DEVICE_ROLE_ROUTER:
+        roleName = OTBR_ROUTER_ROLE_NAME;
+        break;
+    case OT_DEVICE_ROLE_LEADER:
+        roleName = OTBR_LEADER_ROLE_NAME;
+        break;
+    }
+
+    return roleName;
+}
+
+static uint64_t ConvertOpenThreadUint64(const uint8_t aValue[sizeof(uint64_t)])
```

Comment:
This doesn't stop caller from passing a shorter array. Please pass the length explicitly and assert it inside the function. This is to force the caller explicitly provide the length and bugs can be revealed in at least debug mode.